cmake/detect-intrinsics.cmake

```diff
@@ -13,6 +13,74 @@ macro(check_acle_intrinsics)
     set(CMAKE_REQUIRED_FLAGS)
 endmacro()
 
+macro(check_avx512_intrinsics)
+    if(CMAKE_C_COMPILER_ID MATCHES "Intel")
+        if(CMAKE_HOST_UNIX OR APPLE)
+            set(AVX512FLAG "-mavx512f -mavx512dq -mavx512bw -mavx512vl")
+        else()
+            set(AVX512FLAG "/arch:AVX512")
+        endif()
+    elseif(CMAKE_C_COMPILER_ID MATCHES "GNU" OR CMAKE_C_COMPILER_ID MATCHES "Clang")
+        if(NOT NATIVEFLAG)
+            set(AVX512FLAG "-mavx512f -mavx512dq -mavx512bw -mavx512vl")
+        endif()
+    endif()
+    # Check whether compiler supports AVX512 intrinsics
+    set(CMAKE_REQUIRED_FLAGS "${AVX512FLAG}")
+    check_c_source_compile_or_run(
+        "#include <immintrin.h>
+        int main(void) {
+            __m512i x = _mm512_set1_epi16(2);
+            const __m512i y = _mm512_set1_epi16(1);
+            x = _mm512_subs_epu16(x, y);
+            (void)x;
+            return 0;
+        }"
+        HAVE_AVX512_INTRIN
+    )
+    set(CMAKE_REQUIRED_FLAGS)
+endmacro()
+
+macro(check_avx512vnni_intrinsics)
+    if(CMAKE_C_COMPILER_ID MATCHES "Intel")
+        if(CMAKE_HOST_UNIX OR APPLE)
+            # One _probably_ matches the other?
+            set(AVX512VNNIFLAG "-mavx512vnni")
+        else()
+            # Ugh, I have no idea to make visual studio work with this
+            # For now, maybe it's safer to assume it's not supported yet?
+            # I can't find _any_ documentation for this. It _looks_ like
+            # maybe it's not? see: 
+            # https://developercommunity.visualstudio.com/t/internal-compiler-error-when-using-mm256-dpbusd-ep/1596436
+            set(HAVE_AVX512VNNI_INTRIN FALSE)
+        endif()
+    elseif(CMAKE_C_COMPILER_ID MATCHES "GNU" OR CMAKE_C_COMPILER_ID MATCHES "Clang")
+        if(NOT NATIVEFLAG)
+            set(AVX512VNNIFLAG "-mavx512vnni")
+        endif()
+    endif()
+    # Check whether compiler supports AVX512vnni intrinsics
+    set(CMAKE_REQUIRED_FLAGS "${AVX512VNNIFLAG}")
+    # Ugh, I do NOT know what this function's problem is...
```

Comment:
Is this comment necessary? Looks like it was working for me when I tested it.

Reply:
Yeah it's resolved, I think I had some weird stale cache issues.  Sorry, CMake can get tricky sometimes.